tests/vm_test.go

```diff
@@ -1921,6 +1923,72 @@ status:
 			}, 2*time.Minute, 1*time.Second).Should(Succeed())
 		})
 	})
+
+	Context("[Serial] when node becomes unhealthy", Serial, func() {
+		const componentName = "virt-handler"
+		var nodeName string
+		var numberOfVirtHandler int32
+
+		handlerPodGetter := func(virtCli kubecli.KubevirtClient, componentName, namespace, nodeName string) *k8sv1.PodList {
+			pods, err := virtClient.CoreV1().Pods(flags.KubeVirtInstallNamespace).List(context.Background(),
+				metav1.ListOptions{
+					LabelSelector: fmt.Sprintf("kubevirt.io=%s", componentName),
+					FieldSelector: fmt.Sprintf("spec.nodeName=" + nodeName),
+				})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(pods.Items).To(HaveLen(1))
+
+			return pods
+		}
+
+		AfterEach(func() {
+			tests.BlackHolePod(handlerPodGetter, virtClient, componentName, flags.KubeVirtInstallNamespace, nodeName, tests.BlackHoleActionDel)
+			Eventually(tests.GetReadyDaemonSetReplicas(virtClient, componentName, flags.KubeVirtInstallNamespace),
+				30*time.Second, time.Second).Should(BeNumerically("==", numberOfVirtHandler))
+		})
+
+		It("[Serial] the VMs running in that node should be respawned", func() {
+			By("Starting VM")
+			vm := startVM(virtClient, createVM(virtClient, libvmi.NewAlpine()))
+
+			vmi, err := virtClient.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, &k8smetav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			oldUID := vmi.UID
+
+			By("Blocking virt-handler for updating the VMI")
+			nodeName = vmi.Status.NodeName
+			numberOfVirtHandler = tests.GetDesiredDaemonSetReplicas(virtClient, componentName, flags.KubeVirtInstallNamespace)
+			tests.BlackHolePod(handlerPodGetter, virtClient, componentName, flags.KubeVirtInstallNamespace, nodeName, tests.BlackHoleActionAdd)
+			Eventually(tests.GetReadyDaemonSetReplicas(virtClient, componentName, flags.KubeVirtInstallNamespace),
+				120*time.Second, time.Second).Should(BeNumerically("==", numberOfVirtHandler-1))
+
+			pod, err := libvmi.GetPodByVirtualMachineInstance(vmi, vmi.Namespace)
+			Expect(err).ToNot(HaveOccurred())
+
+			By("Simulating lost of the virt-launcher")
+			err = virtClient.CoreV1().Pods(pod.Namespace).Delete(context.Background(), pod.Name, metav1.DeleteOptions{
+				GracePeriodSeconds: ptr.To(int64(0)),
+			})
+			Expect(err).ToNot(HaveOccurred(), "Should delete virt-launcher")
+
+			By("The controller should recreate the launcher pod and move the VMI to phase READY")
+			Eventually(ThisVMI(vmi), 200*time.Second, 2*time.Second).Should(HaveConditionTrue(v1.VirtualMachineInstanceReady))
+
+			By("Waiting for VMI to be running")
+			Eventually(ThisVMI(vmi), 300*time.Second, 1*time.Second).Should(BeRunning())
+
+			By("Waiting for VM to be ready")
+			Eventually(ThisVM(vm), 360*time.Second, 1*time.Second).Should(beReady())
+
+			// This only possible if the VMI has been on FAILED Phase
+			By("Check if the VMI has been recreated")
+			Eventually(func() types.UID {
```

Comment:
I don't think Eventually is still required here. The VMI won't change anymore. Either it was recreated or not at this point.

Reply:
In other words, it needs some time to be updated after the VMI is running.

Reply:
Ok, leaving that for others. Is it expected behavior for the VMI to change its `UID` that late? (VM is already in Ready state)

Reply:
You were right. I was my fault because I wasn't waiting until the VMI was recreated at all, only in the last Eventually statement. Done.